### DISCUSSION TITLE
why many libraries are dropping default export?

### DISCUSSION THREAD

**Discussion by TheMikeyRoss** [pmndrs/zustand#2211] [General] [answered]
I noticed many libraries are dropping default export, why?

**Comment by dai-shi** [comment]
I wonder if @barelyhuman wrote a blog post about it?

I'm sure there has been some issues/discussions regarding it, but generally speaking, we can't simply support various bundlers with default export, because their conventions are different and conflicting.

**Reply by barelyhuman** [reply, marked as answer]
Probably should write one but the basis is the user land issues that come with it and the library developer's having to handle different ways of exporting their API. 


Eg:
If the library exports with `module.exports` then the user will have to do the following if `allowSyntheticDefaultExports` is `false`

```js
import * as libFunction from "lib"
```

There's typescript compiled librares that end up with a `exports.default` and then in ESM you end up having to do this
```js
import libFunctionGroup from "lib"
const fn = libFunctionGroup.libFunction
```

There's more such issues that the users face when libraries are bundled and published without thought and when the library author does put thought in making the library import aesthetic and compatible with all runtimes you end up a large rollup config and package.json `exports` that are pretty fragile, it took us quite a bit of too and fro to get the config and setup we have with our libraries right now

Easier way? 

Only use `exports.functionDef` and `export fnuctionDef` in CJS and ESM respectively because that's just easier for most runtimes to handle
